compiler/rustc_symbol_mangling/src/lib.rs

```diff
@@ -218,35 +218,12 @@ fn compute_symbol_name<'tcx>(
         }
     }
 
-    // Foreign items by default use no mangling for their symbol name. There's a
-    // few exceptions to this rule though:
-    //
-    // * This can be overridden with the `#[link_name]` attribute
-    //
-    // * On the wasm32 targets there is a bug (or feature) in LLD [1] where the
-    //   same-named symbol when imported from different wasm modules will get
-    //   hooked up incorrectly. As a result foreign symbols, on the wasm target,
-    //   with a wasm import module, get mangled. Additionally our codegen will
-    //   deduplicate symbols based purely on the symbol name, but for wasm this
-    //   isn't quite right because the same-named symbol on wasm can come from
-    //   different modules. For these reasons if `#[link(wasm_import_module)]`
-    //   is present we mangle everything on wasm because the demangled form will
-    //   show up in the `wasm-import-name` custom attribute in LLVM IR.
-    //
-    // * `#[rustc_std_internal_symbol]` mangles the symbol name in a special way
-    //   both for exports and imports through foreign items. This is handled above.
-    // [1]: https://bugs.llvm.org/show_bug.cgi?id=44316
-    if tcx.is_foreign_item(def_id)
-        && (!tcx.sess.target.is_like_wasm
-            || !tcx.wasm_import_module_map(def_id.krate).contains_key(&def_id))
-    {
-        if let Some(name) = attrs.link_name {
-            return name.to_string();
-        }
-        return tcx.item_name(def_id).to_string();
+    if let Some(name) = attrs.export_name {
+        // Use provided name
+        return name.to_string();
     }
 
-    if let Some(name) = attrs.export_name {
+    if let Some(name) = attrs.link_name {
```

Comment:
Why are export_name and link_name swapped?

Reply:
they're not swapped, it's a bad diff. We now apply both the export name *and* the link name (see above in the file). export_name is only for foreign items. I thought about merging these two fields. This is likely possible in the future, but rn some logic still happens on export_name that doesn't happen on link_name.

Reply:
Previously it first check the link name for foreign items and then check the export name for non-foreign items. Now it checks export name and then link name irrespective of being a foreign item or not.

By the way it is kinda confusing that we have both export_name and link_name. It seems like only one or the other should be set depending on whether or not the item is a foreign item, but I'm not sure if anything enforces this.

Reply:
that's what I described in the comment above. In theory we can merge these two but that's out of scope for this PR. I'll swap them back around, you're right.